### PR TITLE
[Flink AI Flow] Refactoring API and Implementation

### PR DESCRIPTION
**Pull request by wuchaochen**
## What is the purpose of the change

1. Refactoring Graph, AIGraph and Workflow data structure.
2. Refactoring project structure and ai_flow_context.
3. Refactoring user_define_operation api and Refactoring user_define_control_dependency etc.
4. Refactoring plugins: job plugin, scheduler plugin, blob manager plugin.
5. Add ai_flow_plugins package.
6. Remove useless api(project, configuration etc.)
7. Remove flink_ai_flow, python_ai_flow package

### CODE REVIEW

**Review by SteNicholas — APPROVED** [review]

**Comment by SteNicholas** [comment]
@wuchaochen , please firstly rebase the dev branch and resolve the conflicts.

File: flink-ai-flow/ai_flow/api/project.py [hunk 1]
```diff
@@ -1,344 +0,0 @@
-#
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/project.py:1]: please firstly rebase the dev branch and resolve the conflicts.

File: flink-ai-flow/ai_flow/test/translator/test_base_translator.py [hunk 2]
```diff
@@ -1,405 +0,0 @@
-#
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/test/translator/test_base_translator.py:1]: please firstly rebase the dev branch and resolve the conflicts.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: flink-ai-flow/ai_flow/graph/graph.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+class _IdGenerator(Jsonable):
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.node_type_to_num: Dict[Text, int] = {}
+
+    def generate_id(self, node: BaseNode) -> Text:
+        node_type = type(node).__name__
+        if node_type in self.node_type_to_num:
+            num = self.node_type_to_num[node_type]
```
Comment (bgeng777) [inline flink-ai-flow/ai_flow/graph/graph.py]: Why not directly use  `self.node_type_to_num[node_type] += 1` ?

File: flink-ai-flow/ai_flow/graph/edge.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                 head: Text,
+                 ) -> None:
+        """
+
+        :param tail: the node dependent the other node output id
+        :param head: the node id
+        """
+        super().__init__()
+        if tail is None or head is None:
+            raise Exception('target node id or source node id can not be None!')
```
Comment (bgeng777) [inline flink-ai-flow/ai_flow/graph/edge.py:38]: source -> head?
target -> tail?
In this design, the tail node depends on the head node? Maybe we can update the comment of the variable.

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+from enum import Enum
+from typing import Text
+
+from ai_flow.util.json_utils import Jsonable
+from notification_service.base_notification import UNDEFINED_EVENT_TYPE, DEFAULT_NAMESPACE
+
+from ai_flow.graph.edge import Edge
+
+
+class MetCondition(str, Enum):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:26]: Would `ConditionType` be more accurate?

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    REPEATED = "REPEATED"
+
+
+class MetValueCondition(str, Enum):
+    """
+    EQUAL: the condition that notification service updates a value which equals to the event value
+    UPDATE: the condition that notification service has a update operation on the event key which event
+            value belongs to
+    """
+    EQUAL = "EQUAL"
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:53]: `EQUALS` and `UPDATED`

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                 head: Text,
+                 event_key: Text,
+                 event_value: Text,
+                 event_type: Text = UNDEFINED_EVENT_TYPE,
+                 condition: MetCondition = MetCondition.NECESSARY,
+                 action: TaskAction = TaskAction.START,
+                 life: EventLife = EventLife.ONCE,
+                 value_condition: MetValueCondition = MetValueCondition.EQUAL,
+                 namespace: Text = DEFAULT_NAMESPACE,
+                 sender: Text = None
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:92]: Should we just use a `MetConfig` here?

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    EQUAL: the condition that notification service updates a value which equals to the event value
+    UPDATE: the condition that notification service has a update operation on the event key which event
+            value belongs to
+    """
+    EQUAL = "EQUAL"
+    UPDATE = "UPDATE"
+
+
+class MetConfig(Jsonable):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:57]: ConditionConfig

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                         event_type=self.event_type,
+                         condition=self.condition,
+                         action=self.action,
+                         life=self.life,
+                         value_condition=self.value_condition,
+                         namespace=self.namespace,
+                         sender=self.sender)
+
+
+class AIFlowInnerEventType(object):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:117]: `AIFlowInternalEventType`

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    JOB_STATUS_CHANGED = "JOB_STATUS_CHANGED"
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:118]: Can we add docs to these events?

File: flink-ai-flow/ai_flow/workflow/job.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+# under the License.
+#
+from typing import Text
+from ai_flow.graph.node import BaseNode
+from ai_flow.workflow.job_config import JobConfig
+
+
+class Job(BaseNode):
+    """
+    A Job contains the common information of a ai flow job. Users can implement custom jobs by adding other
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/job.py:26]: It took me quite a while to read this doc yet still do not understand what exactly this means....

File: flink-ai-flow/ai_flow/workflow/job_config.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+class JobConfig(Jsonable):
+    """
+    Base class for job config. It is used to set the basic job config.
+
+    """
+
+    def __init__(self,
+                 job_name: Text = None,
+                 job_type: Text = None,
+                 exec_mode: Optional[ExecutionMode] = ExecutionMode.BATCH,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/job_config.py:33]: Do we still need this?

File: flink-ai-flow/ai_flow/workflow/job_config.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        self.job_name: Text = job_name
+        self.job_type = job_type
+        self.exec_mode: Optional[ExecutionMode] = exec_mode
+        if properties is None:
+            self.properties: Dict[Text, Jsonable] = {}
+        else:
+            self.properties: Dict[Text, Jsonable] = properties
+
+    @staticmethod
+    def from_dict(job_name: Text, data: Dict) -> 'JobConfig':
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/job_config.py:52]: Will users call this method? Should it be a "private" method with __ as prefix? 

Also, its seems that the `job_name` is a part of the `JobConfig`, but not loaded from the `data`. That means the return value of `to_dict()` might not be restored by the `from_dict()` method. Is there a reason we do this?

File: flink-ai-flow/ai_flow/workflow/job_state.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+class JobState(object):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/job_state.py]: We probably need docs for each of these states.

File: flink-ai-flow/ai_flow/workflow/periodic_config.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class PeriodicConfig(Jsonable):
+    """
+    Set period semantics for job. Jobs with periodic semantics will restart at regular intervals.
+    """
+    def __init__(self, periodic_type: Text,
+                 args: Dict[Text, Jsonable] = None) -> None:
+        """
+        :param periodic_type: ``interval`` or ``cron``
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/periodic_config.py]: Why is the type not an enum?

File: flink-ai-flow/ai_flow/workflow/workflow.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+    @property
+    def control_edges(self)->Dict[Text, List[ControlEdge]]:
+        return self.edges
+
+    def add_job(self, job: Job):
+        self.nodes[job.job_config.job_name] = job
+
+    def add_edges(self, job_name: Text, dependencies: List[ControlEdge]):
+        self.edges[job_name] = dependencies
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/workflow.py:50]: Should we copy the dependencies to a new List instead of using the one passed in? This will avoid the potential risk that the argument is modified later.

File: flink-ai-flow/ai_flow/workflow/workflow.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    def add_edge(self, job_name: Text, edge: ControlEdge):
+        if job_name not in self.edges:
+            self.edges[job_name] = []
+        self.edges[job_name].append(edge)
+
+
+class WorkflowPropertyKeys(object):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/workflow.py:58]: Can we add docs to the property keys?

**Review by becketqin — COMMENTED** [review]
@wuchaochen Thanks for the patch. I took a quick pass on the first 3 commits and left some comments. Will continue reviewing the patch.

File: flink-ai-flow/ai_flow/plugin_interface/scheduler_interface.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+# under the License.
+from abc import ABC, abstractmethod
+from typing import Dict, Text, List, Optional
+from ai_flow.common.configuration import AIFlowConfiguration
+from ai_flow.util import json_utils
+from ai_flow.workflow.workflow import Workflow
+from ai_flow.project.project_description import ProjectDesc
+
+
+class WorkflowInfo(json_utils.Jsonable):
```
Comment (Sxnan) [inline flink-ai-flow/ai_flow/plugin_interface/scheduler_interface.py:25]: Can we get rid of the setters in *Info class? I think those fields should be immutable.

File: flink-ai-flow/ai_flow/plugin_interface/scheduler_interface.py [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+    def properties(self, value):
+        self._properties = value
+
+    def __str__(self) -> str:
+        return json_utils.dumps(self)
+
+
+class SchedulerConfig(AIFlowConfiguration):
+
+    def repository(self):
```
Comment (Sxnan) [inline flink-ai-flow/ai_flow/plugin_interface/scheduler_interface.py:216]: What is the repository? How is it used by the scheduler?

File: flink-ai-flow/ai_flow/runtime/job_runtime_env.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        return os.path.join(self.working_dir, '{}.python'.format(self.workflow_name))
+
+    def save_workflow_name(self):
+        if self._workflow_name is None:
+            return
+        file_path = os.path.join(self.working_dir, 'workflow_name')
+        if os.path.exists(file_path):
+            os.remove(file_path)
+        with open(file_path, 'wb') as fp:
+            fp.write(serialization_utils.serialize(self._workflow_name))
```
Comment (Sxnan) [inline flink-ai-flow/ai_flow/runtime/job_runtime_env.py:100]: Do we have to serialize a workflow_name and job_name?

**Review by becketqin — COMMENTED** [review]
@wuchaochen I left a few more comments. It looks that quite some docs are missing. Let's fix that as the docs make sure the code is readable and understandable to the contributors.

**Review by jiangxin369 — APPROVED** [review]

File: flink-ai-flow/ai_flow/runtime/job_runtime_env.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import os
+from typing import Text
+from ai_flow.util import serialization_utils
+
+
+class JobRuntimeEnv(object):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/runtime/job_runtime_env.py:22]: This class seems similar to the ProjectDesc, as they both describe the project structure in some sense. Can we somehow consolidate these two classes a little bit?

File: flink-ai-flow/ai_flow/test/plugin_interface/test_blob_manager.py [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    def upload_blob(self, workflow_id: Text, prj_pkg_path: Text) -> Text:
+        return 'upload'
+
+    def download_blob(self, workflow_id, remote_path: Text, local_path: Text = None) -> Text:
+        return 'download'
+
+
+class TestBlobManager(unittest.TestCase):
+
+    def test_custom_blob_manager(self):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/test/plugin_interface/test_blob_manager.py:37]: Is the purpose of this test to test the blob_manager_factory? If so can we just rename the method to `test_blob_manager_factory()`, so it is clear that we are not testing the `MockBlockManager`.

File: flink-ai-flow/ai_flow/test/plugin_interface/test_job_plugin_interface.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+    def stop_job(self, job_handler: JobHandler, job_context: JobExecutionContext):
+        pass
+
+    def cleanup_job(self, job_handler: JobHandler, job_context: JobExecutionContext):
+        pass
+
+
+class TestJobPlugin(unittest.TestCase):
+
+    def test_job_plugin(self):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/test/plugin_interface/test_job_plugin_interface.py:66]: test_register_job_plugin()

File: flink-ai-flow/ai_flow/test/runtime/test_job_runtime_context.py [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+from ai_flow.context.job_context import job_config, current_job_name
+from ai_flow.context.project_context import project_config
+from ai_flow.context.workflow_context import workflow_config
+from ai_flow.runtime.job_runtime_env import JobRuntimeEnv
+from ai_flow.runtime.job_runtime_context import init_job_runtime_context
+
+
+class TestJobRuntimeContext(unittest.TestCase):
+
+    def test_job_runtime_env(self):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/test/runtime/test_job_runtime_context.py:30]: `test_init_job_runtime_context()`. Usually the unit test class name is `TestMyClassName`, while each testing method is test a method in the class or some particular scenario. So the name of the testing method will be `testMyMethodName` or `testMyScenarioName`.

File: flink-ai-flow/ai_flow/api/ops.py [hunk 25]
```diff
@@ -0,0 +1,582 @@
...
+from ai_flow.ai_graph.ai_graph import default_graph, add_ai_node_to_graph
+from ai_flow.meta.dataset_meta import DatasetMeta
+from ai_flow.meta.model_meta import ModelMeta, ModelVersionMeta
+from ai_flow.context.project_context import project_description
+from ai_flow.context.workflow_context import workflow_config
+
+
+def read_dataset(dataset_info: Union[DatasetMeta, Text, int],
+                 executor=None,
+                 **kwargs) -> Channel:
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:37]: It seems that this varargs is not necessary. The parameters here can be categorized to three different types:
1. Parameters to AI Flow - such as dataset_info here, so that AI Flow can help do the bookkeeping like lineage.
2. Parameters to the `Executor` - in this case, the parameters can be passed to the constructor of the `Executor`.
3. Parameters to the job generator - Such parameters passed in here are supposed to be used by the job generator to generate the job for this particular Executor. 

It looks that the varargs here are for the 3rd use case. Arguably, we can also put these per `Executor` parameters in the `Executor` parameters. This essentially lets the `Executor` plugin pass these parameters to the job generator and thus makes the `read_dataset()` API simpler. The same design principle could be applied to the other methods.

File: flink-ai-flow/ai_flow/api/ops.py [hunk 26]
```diff
@@ -0,0 +1,582 @@
...
+        dataset_meta = get_ai_flow_client().get_dataset_by_id(dataset_info)
+
+    return user_define_operation(dataset_info=dataset_meta,
+                                 executor=executor,
+                                 output_num=1,
+                                 operation_type='read_dataset',
+                                 **kwargs)
+
+
+def write_dataset(input_data: Channel,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:62]: Maybe just `input`

File: flink-ai-flow/ai_flow/api/ops.py [hunk 27]
```diff
@@ -0,0 +1,582 @@
...
+                                 executor=executor,
+                                 output_num=0,
+                                 operation_type='write_dataset',
+                                 **kwargs)
+
+
+def transform(input_data: Union[Channel, List[Channel]],
+              executor,
+              output_num=1,
+              name: Text = None,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:96]: Do `read_dataset` and `write_dataset` also need the name field?

File: flink-ai-flow/ai_flow/api/ops.py [hunk 28]
```diff
@@ -0,0 +1,582 @@
...
+    """
+    return user_define_operation(input_data=input_data,
+                                 name=name,
+                                 executor=executor,
+                                 output_num=output_num,
+                                 operation_type='transform',
+                                 **kwargs)
+
+
+def train(input_data: Union[Channel, List[Channel]],
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:117]: Would it be better to have a more meaningful name for the parameters? e.g. `training_samples`. Similarly, for the `Executor`, it could be `training_executor`. This also applies to the other ops.

File: flink-ai-flow/ai_flow/api/ops.py [hunk 29]
```diff
@@ -0,0 +1,582 @@
...
+                                 name=name,
+                                 model_info=model_meta,
+                                 executor=executor,
+                                 model_version_info=model_version_meta,
+                                 output_num=output_num,
+                                 operation_type='predict',
+                                 **kwargs)
+
+
+def evaluate(input_data: Union[Channel, List[Channel]],
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:219]: Do we expect the evaluator executor to send events after the evaluation to expose the evaluation result? It might be useful to give a typical behavior of the ops.

File: flink-ai-flow/ai_flow/api/ops.py [hunk 30]
```diff
@@ -0,0 +1,582 @@
...
+    return user_define_operation(input_data=input_data,
+                                 name=name,
+                                 model_info=model_meta,
+                                 executor=executor,
+                                 output_num=output_num,
+                                 operation_type='evaluate',
+                                 **kwargs)
+
+
+def dataset_validate(input_data: Channel,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:254]: So we do not expect users to override the `output_num` here?

File: flink-ai-flow/ai_flow/api/ops.py [hunk 31]
```diff
@@ -0,0 +1,582 @@
...
+    return user_define_operation(input_data=input_data,
+                                 executor=executor,
+                                 name=name,
+                                 output_num=0,
+                                 operation_type='dataset_validate',
+                                 **kwargs
+                                 )
+
+
+def model_validate(input_data: Union[Channel, List[Channel]],
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:276]: Can we also provide a typical / expected behavior here to report the chosen model version?

File: flink-ai-flow/ai_flow/api/ops.py [hunk 32]
```diff
@@ -0,0 +1,582 @@
...
+                    event_key: Text,
+                    event_value: Text,
+                    event_type: Text = UNDEFINED_EVENT_TYPE,
+                    condition: MetCondition = MetCondition.NECESSARY,
+                    action: TaskAction = TaskAction.START,
+                    life: EventLife = EventLife.ONCE,
+                    value_condition: MetValueCondition = MetValueCondition.EQUAL,
+                    namespace: Text = DEFAULT_NAMESPACE,
+                    sender: Text = None
+                    ):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:439]: The parameter order here seems a little unclear, can we adjust the parameter order a little bit to put the `condition`, `value_condition` and `action` to the last?

File: flink-ai-flow/ai_flow/api/ops.py [hunk 33]
```diff
@@ -0,0 +1,582 @@
...
+                               action=action,
+                               life=life,
+                               value_condition=value_condition,
+                               namespace=namespace,
+                               sender=sender
+                               )
+    default_graph().add_edge(job_name, control_edge)
+
+
+def action_on_model_version_event(job_name: Text,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/api/ops.py:472]: action_on_model_version_event, action_on_dataset_event are essentially built-in control dependencies based on some predefined events. We need to tell a complete story about those events, otherwise uses might not understand where are those events from.

File: flink-ai-flow/ai_flow/ai_graph/data_edge.py [hunk 34]
```diff
@@ -15,16 +14,30 @@
...
+                 data_config: Dict[Text, Jsonable] = None) -> None:
+        super().__init__(head=head, tail=tail)
+        self.port = port
+        self.data_config = data_config
+
+    def __eq__(self, o: object) -> bool:
+        if isinstance(o, DataEdge):
+            return self.head == o.head \
+                   and self.tail == o.tail \
+                   and self.port == o.port
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/ai_graph/data_edge.py:38]:
```suggestion
                   and self.port == o.port \
                   and self.data_config == self.data_config
```

File: flink-ai-flow/ai_flow/api/ops.py [hunk 35]
```diff
@@ -486,57 +377,39 @@ def push_model(model_info: Union[ModelMeta, Text, int],
...
+        **kwargs) -> Union[NoneChannel, Channel, Tuple[Channel]]:
     """
     User defined operator.
 
+    :param operation_type: The type of the operation.
     :param executor: The user defined function in operator. User can write their own logic here.
-    :param input_data_list: List of input data. It contains multiple channels from the operators which generate data.
-    :param exec_args: The properties of read dataset, there are batch properties, stream properties and
-                      common properties respectively.
+    :param input_data: List of input data. It contains multiple channels from the operators which generate data.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:400]:
```suggestion
    :param input_data: List of Channel. It contains multiple channels from the operators which generate data.
```

File: flink-ai-flow/ai_flow/api/ops.py [hunk 36]
```diff
@@ -553,151 +426,159 @@ def user_define_operation(
...
-                                     model_version_event_type,
-                                     dependency: Channel,
-                                     namespace: Text = DEFAULT_NAMESPACE
-                                     ) -> None:
+       Add user defined control logic.
+       :param job_name: The job name identify the job.
+       :param namespace: the project name
+       :param event_key: The key of the event.
+       :param event_value: The value of the event.
+       :param event_type: The Name of the event.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:446]:
```suggestion
       :param event_type: The type of the event.
```

File: flink-ai-flow/ai_flow/api/ops.py [hunk 37]
```diff
@@ -553,151 +426,159 @@ def user_define_operation(
...
-
-def model_version_control_dependency(src: Channel,
-                                     model_name: Text,
-                                     model_version_event_type,
-                                     dependency: Channel,
-                                     namespace: Text = DEFAULT_NAMESPACE
-                                     ) -> None:
+       Add user defined control logic.
+       :param job_name: The job name identify the job.
+       :param namespace: the project name
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:443]:
```suggestion
       :param namespace: The namespace of the event, which value uses the project name.
```

File: flink-ai-flow/ai_flow/api/ops.py [hunk 38]
```diff
@@ -553,151 +426,159 @@ def user_define_operation(
...
+       :param event_type: The Name of the event.
+       :param condition: The event condition. Sufficient or Necessary.
+       :param action: The action act on the src channel. Start or Restart.
+       :param life: The life of the event. Once or Repeated.
+       :param value_condition: The event value condition. Equal or Update. Equal means the src channel will start or
+                               restart only when in the condition that the notification service updates a value which
+                               equals to the event value under the specific event key, while update means src channel
+                               will start or restart when in the the condition that the notification service has a update
+                               operation on the event key which event value belongs to.
+       :param sender: The event sender identity. If sender is None, the sender will be dependency.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:455]:
```suggestion
       :param sender: The event sender identity, which value uses the name of the job. If sender is None, the sender will be dependency.
```

File: flink-ai-flow/ai_flow/api/ops.py [hunk 39]
```diff
@@ -553,151 +426,159 @@ def user_define_operation(
...
+                               value_condition=value_condition,
+                               namespace=namespace,
+                               sender=sender
+                               )
+    default_graph().add_edge(job_name, control_edge)
+
+
+def action_on_model_version_event(job_name: Text,
+                                  model_name: Text,
+                                  model_version_event_type: Text,
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:474]: What does `model_version_event_type` mean? Do you mean the `model_version_event_type` represent the `status` of the model version? What's more, there's no comment for this parameter.

File: flink-ai-flow/ai_flow/api/ops.py [hunk 40]
```diff
@@ -553,151 +426,159 @@ def user_define_operation(
...
+                    event_type=AIFlowInnerEventType.DATASET_CHANGED,
+                    action=action,
+                    namespace=namespace,
+                    sender=ANY_CONDITION
+                    )
+
+
+def action_on_state(job_name: Text,
+                    upstream_job_name: Text,
+                    upstream_job_state: Text,
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:527]: Does the type of `upstream_job_state` need to use Enum? IMO, the users don't know how to set the value of this parameter if here doesn't use Enum.

File: flink-ai-flow/ai_flow/api/ops.py [hunk 41]
```diff
@@ -553,151 +426,159 @@ def user_define_operation(
...
+                    action=action,
+                    namespace=project_description().project_name,
+                    condition=MetCondition.SUFFICIENT
+                    )
+
+
+def run_after(job_name: Text,
+              upstream_job_name: Text) -> None:
+    """
+    Job run after upstream job success.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/ops.py:551]:
```suggestion
    Job run after upstream job finishes.
```

File: flink-ai-flow/ai_flow/api/workflow_operation.py [hunk 42]
```diff
@@ -78,25 +86,26 @@ def submit_workflow(workflow_name: Text = None,
...
-    _set_entry_module_path(workflow, entry_module_path)
-    _upload_project_package(workflow)
     return proto_to_workflow(get_ai_flow_client()
                              .submit_workflow_to_scheduler(namespace=namespace,
                                                            workflow_json=json_utils.dumps(workflow),
                                                            workflow_name=workflow_name,
                                                            args=args))
 
 
+def apply_full_info_to_workflow(entry_module_path, workflow):
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/api/workflow_operation.py:102]: Should the `apply_full_info_to_workflow` method be private? Users don't need to invoke the method.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py [hunk 43]
```diff
@@ -0,0 +1,328 @@
...
+        return None
+
+
+class FlinkJobPlugin(AbstractJobPlugin, ABC):
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    @classmethod
+    def build_run_graph(cls, sub_graph: AISubGraph) -> RunGraph:
```
Comment (jiangxin369) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:143]: This method is same as `build_run_graph` in `PythonJobPlugin`, why not make it a common one

**Review by becketqin — COMMENTED** [review]
Thanks for the patch. It looks that the commit history of the PR is messed up. Please clean up history to make the reviewers' life easier. Here are my suggestions:
1. Keep the commits of API changes here. Note that there seems to be a few "Merge" commits which removes some code in the previous commits and the same code are again added back in later commits. That needs to be fixed.
2. Have separate PRs for Python Job Plugin and Flink JobPlugin respectively.

File: flink-ai-flow/ai_flow/runtime/job_runtime_env.py [hunk 44]
```diff
@@ -100,11 +126,11 @@ def save_workflow_name(self):
         with open(file_path, 'wb') as fp:
             fp.write(serialization_utils.serialize(self._workflow_name))
 
-    def save_job_name(self):
-        if self._job_name is None:
+    def save_job_execution_info(self):
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/runtime/job_runtime_env.py:129]: Who will invoke this method?

File: flink-ai-flow/ai_flow/scheduler/scheduler_service.py [hunk 45]
```diff
@@ -123,30 +147,12 @@ def resumeWorkflowScheduling(self, request, context):
...
-            if workflow_info is None:
-                return WorkflowInfoResponse(
-                    result=ResultProto(status=StatusProto.ERROR,
-                                       error_message='{} do not exist!'.format(rq.workflow_name)))
-            return WorkflowInfoResponse(result=ResultProto(status=StatusProto.OK),
-                                        workflow=workflow_to_proto(workflow_info))
-        except Exception as err:
-            return WorkflowInfoResponse(result=ResultProto(status=StatusProto.ERROR,
-                                                           error_message=traceback.format_exc()))
+        return WorkflowInfoResponse(result=ResultProto(status=StatusProto.ERROR,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/scheduler/scheduler_service.py:150]: Just to clarify, we are going to call metadata client to get, list and delete workflow, right? Will this be done in a separate PR?

File: flink-ai-flow/ai_flow/workflow/control_edge.py [hunk 46]
```diff
@@ -59,12 +59,12 @@ def __init__(self,
                  event_key: Text,
                  event_value: Text,
                  event_type: Text = UNDEFINED_EVENT_TYPE,
+                 namespace: Text = DEFAULT_NAMESPACE,
+                 sender: Text = None,
                  condition: ConditionType = ConditionType.NECESSARY,
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/control_edge.py:64]: Can we change the name to `ConditionType`?

File: flink-ai-flow/ai_flow/workflow/status.py [hunk 47]
```diff
@@ -17,11 +17,11 @@
...
-class State(str, Enum):
-    INIT = 'INIT'
-    STARTING = 'STARTING'
-    RUNNING = 'RUNNING'
-    FINISHED = 'FINISHED'
-    FAILED = 'FAILED'
-    KILLING = 'KILLING'
-    KILLED = 'KILLED'
+class Status(str, Enum):
+    INIT = 'INIT'  # The scheduler create the job/workflow and do not scheduling it.
```
Comment (becketqin) [inline flink-ai-flow/ai_flow/workflow/status.py:21]: We need to clearly state the transition condition of the states.

**Review by becketqin — COMMENTED** [review]
The FlinkJobGenerator / PythonJobGenerator should be separated out to two different PRs respectively.

**Comment by becketqin** [comment]
@wuchaochen Thanks for updating the patch. Make a quick pass. Please reorg the commits to remove the back-and-forth change. Also, I would suggest to remove the implementation of job plugins and scheduler plugins to separate PRs.

File: flink-ai-flow/ai_flow_plugins/scheduler_plugins/airflow/ai_flow_operator.py [hunk 48]
```diff
@@ -0,0 +1,105 @@
...
+        self.job: Job = job
+        self.workflow: Workflow = workflow
+        plugins = self.workflow.properties.get(WorkflowPropertyKeys.JOB_PLUGINS)
+        module, name = plugins.get(self.job.job_config.job_type)
+        class_object = import_string('{}.{}'.format(module, name))
+        self.job_controller: JobController = class_object()
+        self.job_handler: JobHandle = None
+        self.job_runtime_env: JobRuntimeEnv = None
+
+    def context_to_job_info(self, project_name: Text, context: Any) -> JobExecutionInfo:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/scheduler_plugins/airflow/ai_flow_operator.py:50]: The `context_to_job_info` method could not bind with the context from Airflow. The AIFlowOperator should be not about the scheduler.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: flink-ai-flow/ai_flow_plugins/scheduler_plugins/airflow/airflow_scheduler.py [hunk 49]
```diff
@@ -319,15 +292,17 @@ def get_job(self, job_name: Text, execution_id: Text) -> Optional[JobInfo]:
...
             else:
                 project_name, workflow_name = self.dag_id_to_namespace_workflow(dag_run.dag_id)
-                return JobInfo(job_name=job_name,
-                               state=self.airflow_state_to_state(task.state),
-                               workflow_execution
-                               =WorkflowExecutionInfo(workflow_info=WorkflowInfo(namespace=project_name,
-                                                                                 workflow_name=workflow_name),
-                                                      execution_id=dag_run.run_id,
-                                                      state=self.airflow_state_to_state(dag_run.state)))
+                return [JobExecutionInfo(job_name=job_name,
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/scheduler_plugins/airflow/airflow_scheduler.py:295]: The `job_execution_id` of the `JobExecutionInfo` doesn't set the value with task.try_number.

File: flink-ai-flow/ai_flow_plugins/scheduler_plugins/airflow/airflow_scheduler.py [hunk 50]
```diff
@@ -340,12 +315,14 @@ def list_jobs(self, execution_id: Text) -> List[JobInfo]:
...
                 project_name, workflow_name = self.dag_id_to_namespace_workflow(dag_run.dag_id)
                 for task in task_list:
-                    job = JobInfo(job_name=task.task_id,
-                                  state=self.airflow_state_to_state(task.state),
-                                  workflow_execution
-                                  =WorkflowExecutionInfo(workflow_info=WorkflowInfo(namespace=project_name,
-                                                                                    workflow_name=workflow_name),
-                                                         execution_id=dag_run.run_id,
-                                                         state=self.airflow_state_to_state(dag_run.state)))
+                    job = JobExecutionInfo(job_name=task.task_id,
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/scheduler_plugins/airflow/airflow_scheduler.py:318]: The `job_execution_id` of the `JobExecutionInfo` doesn't set the value with task.try_number.

**Review by SteNicholas — CHANGES_REQUESTED** [review]
